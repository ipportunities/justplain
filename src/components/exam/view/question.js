import React from 'react'
import { MultipleChoiceQuestion, OpenQuestion } from './questionInputTypes'

export const ViewQuestion = ({ question, coachView, setAnswer, setPoints, setComment }) => {
  let inputTypeMap = {
    'MULTIPLE_CHOICE': MultipleChoiceQuestion,
    'OPEN': OpenQuestion
  }

  return <div className={ coachView ? 'col-8 offset-2' : 'col' }>
    <div className="row pt-2 pb-3">
      <div className="col">
        <div className="row pb-2">
          <div className="col-12">
            { question.question }
          </div>
        </div>
        <div className="row">
          <div className="col-12">
            <label className="mt-2">{ coachView ? '(Suggested) correct answer' : 'Answer' }</label>
            { question.answers ? React.createElement(inputTypeMap[question.type], {
              coachView,
              question,
              setAnswer
            }) : null }
          </div>
        </div>
        { question.result && coachView ? <>
          <div className="row">
            <div className="col-12">
              <label className="mt-2">Given answer</label>
              { question.answers ? React.createElement(inputTypeMap[question.type], {
                coachView,
                answerView: true,
                question,
                setAnswer
              }) : null }
            </div>
          </div>
          <div className="row">
            <div className="col-12">
              <label className="mt-2">Comments</label>
              <textarea className="form-control p-2" value={ question.result.comment } rows={ 4 }
                        onChange={ setComment ? e => setComment(e.target.value) : () => {} }
                        readOnly={ !setComment } />
            </div>
          </div>
        </> : null
        }
      </div>
      { coachView &&
      <div className="col-3">
        <label>Type</label>
        <input type="text" className="form-control pl-1 pr-1"
               value={ (question.type.charAt(0).toUpperCase() + question.type.slice(1).toLowerCase()).replace('_', ' ') } />
        <label>Difficulty</label>
        <input type="text" className="form-control pl-1 pr-1"
               value={ question.difficulty.charAt(0).toUpperCase() + question.difficulty.slice(1).toLowerCase() } />
        <label>Subject</label>
        <input type="text" className="form-control pl-1 pr-1" value={ question.subject } />
        <label>{ `${ question.type === 'OPEN' ? 'Maximum achievable points' : 'Achievable points' }` }</label>
        <input type="text" className="form-control pl-1 pr-1" value={ question.maxPoints } />
        { question.result ? <>
          <label>Points achieved</label>
          <input type="number" className="form-control pl-1 pr-1" value={ question.result.points }
                 readOnly={ !setPoints }
                 onChange={ setPoints ? e => setPoints(e.target.value) : () => {} } min="0"
                 max={ question.maxPoints } />
        </> : null }
      </div>
      }
    </div>
  </div>
}
