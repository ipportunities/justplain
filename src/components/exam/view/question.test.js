import React from 'react'
import { shallow } from 'enzyme'
import { ViewQuestion } from './question'

const openQuestionMock = {
  id: 0,
  question: 'Question 0 - OPEN ?',
  answers: [
    { answer: 'A', correct: true, order: 0 }
  ],
  result: {
    answer: 'A',
  },
  type: 'OPEN',
  difficulty: 'EASY',
  subject: 'A',
  maxPoints: 10
}
const multipleChoiceQuestionMock = {
  id: 1,
  question: 'Question 1 - MULTIPLE_CHOICE ?',
  answers: [
    { answer: 'A', correct: true, order: 0 },
    { answer: 'B', correct: false, order: 1 },
    { answer: 'C', correct: false, order: 2 }
  ],
  result: {
    answer: 'B',
  },
  type: 'MULTIPLE_CHOICE',
  difficulty: 'MEDIUM',
  subject: 'B',
  maxPoints: 5
}

describe('View Question Coach', () => {
  it('should render correctly open answer', () => {
    const component = shallow(<ViewQuestion question={ openQuestionMock } coachView={ true } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly multiple choice answer', () => {
    const component = shallow(<ViewQuestion question={ multipleChoiceQuestionMock } coachView={ true } />)
    expect(component).toMatchSnapshot()
  })
})

describe('View Question Student', () => {
  it('should render correctly open answer', () => {
    const component = shallow(<ViewQuestion question={ openQuestionMock } setAnswer={ jest.fn() } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly open answer with given answer', () => {
    const q = openQuestionMock
    q.givenAnswer = 'A'
    const component = shallow(<ViewQuestion question={ q } setAnswer={ jest.fn() } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly multiple choice answer', () => {
    const component = shallow(<ViewQuestion question={ multipleChoiceQuestionMock } setAnswer={ jest.fn() } />)
    expect(component).toMatchSnapshot()
  })
  it('should render correctly multiple choice answer with given answer', () => {
    const q = openQuestionMock
    q.givenAnswer = 1
    const component = shallow(<ViewQuestion question={ multipleChoiceQuestionMock } setAnswer={ jest.fn() } />)
    expect(component).toMatchSnapshot()
  })
})
