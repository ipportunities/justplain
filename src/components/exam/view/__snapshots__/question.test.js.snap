// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Question Coach should render correctly multiple choice answer 1`] = `
<div
  className="col-8 offset-2"
>
  <div
    className="row pt-2 pb-3"
  >
    <div
      className="col"
    >
      <div
        className="row pb-2"
      >
        <div
          className="col-12"
        >
          Question 1 - MULTIPLE_CHOICE ?
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <label
            className="mt-2"
          >
            Suggested correct answer
          </label>
          <div
            className="form-check"
            key="0"
          >
            <input
              checked={1}
              className="form-check-input"
              id="checkbox-answer-0"
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-0"
            >
              A
            </label>
          </div>
          <div
            className="form-check"
            key="1"
          >
            <input
              checked={0}
              className="form-check-input"
              id="checkbox-answer-1"
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-1"
            >
              B
            </label>
          </div>
          <div
            className="form-check"
            key="2"
          >
            <input
              checked={0}
              className="form-check-input"
              id="checkbox-answer-2"
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-2"
            >
              C
            </label>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <label
            className="mt-2"
          >
            Given answer
          </label>
          <div
            className="form-check"
            key="0"
          >
            <input
              checked={false}
              className="form-check-input"
              id="checkbox-answer-0"
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-0"
            >
              A
            </label>
          </div>
          <div
            className="form-check"
            key="1"
          >
            <input
              checked={false}
              className="form-check-input"
              id="checkbox-answer-1"
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-1"
            >
              B
            </label>
          </div>
          <div
            className="form-check"
            key="2"
          >
            <input
              checked={false}
              className="form-check-input"
              id="checkbox-answer-2"
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-2"
            >
              C
            </label>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <label
            className="mt-2"
          >
            Comments
          </label>
          <textarea
            className="form-control p-2"
            onChange={[Function]}
            readOnly={true}
            rows={4}
          />
        </div>
      </div>
    </div>
    <div
      className="col-3"
    >
      <label>
        Type
      </label>
      <input
        className="form-control pl-1 pr-1"
        type="text"
        value="Multiple choice"
      />
      <label>
        Difficulty
      </label>
      <input
        className="form-control pl-1 pr-1"
        type="text"
        value="Medium"
      />
      <label>
        Subject
      </label>
      <input
        className="form-control pl-1 pr-1"
        type="text"
        value="B"
      />
      <label>
        Achievable points
      </label>
      <input
        className="form-control pl-1 pr-1"
        type="text"
        value={5}
      />
      <label>
        Points achieved
      </label>
      <input
        className="form-control pl-1 pr-1"
        max={5}
        min="0"
        onChange={[Function]}
        readOnly={true}
        type="number"
      />
    </div>
  </div>
</div>
`;

exports[`View Question Coach should render correctly open answer 1`] = `
<div
  className="col-8 offset-2"
>
  <div
    className="row pt-2 pb-3"
  >
    <div
      className="col"
    >
      <div
        className="row pb-2"
      >
        <div
          className="col-12"
        >
          Question 0 - OPEN ?
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <label
            className="mt-2"
          >
            Suggested correct answer
          </label>
          <div>
            <textarea
              className="form-control p-2"
              onChange={[Function]}
              rows={4}
              value="A"
            />
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <label
            className="mt-2"
          >
            Given answer
          </label>
          <div>
            <textarea
              className="form-control p-2"
              rows={4}
            />
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <label
            className="mt-2"
          >
            Comments
          </label>
          <textarea
            className="form-control p-2"
            onChange={[Function]}
            readOnly={true}
            rows={4}
          />
        </div>
      </div>
    </div>
    <div
      className="col-3"
    >
      <label>
        Type
      </label>
      <input
        className="form-control pl-1 pr-1"
        type="text"
        value="Open"
      />
      <label>
        Difficulty
      </label>
      <input
        className="form-control pl-1 pr-1"
        type="text"
        value="Easy"
      />
      <label>
        Subject
      </label>
      <input
        className="form-control pl-1 pr-1"
        type="text"
        value="A"
      />
      <label>
        Maximum achievable points
      </label>
      <input
        className="form-control pl-1 pr-1"
        type="text"
        value={10}
      />
      <label>
        Points achieved
      </label>
      <input
        className="form-control pl-1 pr-1"
        max={10}
        min="0"
        onChange={[Function]}
        readOnly={true}
        type="number"
      />
    </div>
  </div>
</div>
`;

exports[`View Question Student should render correctly multiple choice answer 1`] = `
<div
  className="col"
>
  <div
    className="row pt-2 pb-3"
  >
    <div
      className="col"
    >
      <div
        className="row pb-2"
      >
        <div
          className="col-12"
        >
          Question 1 - MULTIPLE_CHOICE ?
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <label
            className="mt-2"
          >
            Answer
          </label>
          <div
            className="form-check"
            key="0"
          >
            <input
              checked={false}
              className="form-check-input"
              id="checkbox-answer-0"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-0"
            >
              A
            </label>
          </div>
          <div
            className="form-check"
            key="1"
          >
            <input
              checked={false}
              className="form-check-input"
              id="checkbox-answer-1"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-1"
            >
              B
            </label>
          </div>
          <div
            className="form-check"
            key="2"
          >
            <input
              checked={false}
              className="form-check-input"
              id="checkbox-answer-2"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-2"
            >
              C
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`View Question Student should render correctly multiple choice answer with given answer 1`] = `
<div
  className="col"
>
  <div
    className="row pt-2 pb-3"
  >
    <div
      className="col"
    >
      <div
        className="row pb-2"
      >
        <div
          className="col-12"
        >
          Question 1 - MULTIPLE_CHOICE ?
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <label
            className="mt-2"
          >
            Answer
          </label>
          <div
            className="form-check"
            key="0"
          >
            <input
              checked={false}
              className="form-check-input"
              id="checkbox-answer-0"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-0"
            >
              A
            </label>
          </div>
          <div
            className="form-check"
            key="1"
          >
            <input
              checked={false}
              className="form-check-input"
              id="checkbox-answer-1"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-1"
            >
              B
            </label>
          </div>
          <div
            className="form-check"
            key="2"
          >
            <input
              checked={false}
              className="form-check-input"
              id="checkbox-answer-2"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="form-check-label"
              htmlFor="checkbox-answer-2"
            >
              C
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`View Question Student should render correctly open answer 1`] = `
<div
  className="col"
>
  <div
    className="row pt-2 pb-3"
  >
    <div
      className="col"
    >
      <div
        className="row pb-2"
      >
        <div
          className="col-12"
        >
          Question 0 - OPEN ?
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <label
            className="mt-2"
          >
            Answer
          </label>
          <div>
            <textarea
              className="form-control p-2"
              onChange={[Function]}
              rows={4}
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`View Question Student should render correctly open answer with given answer 1`] = `
<div
  className="col"
>
  <div
    className="row pt-2 pb-3"
  >
    <div
      className="col"
    >
      <div
        className="row pb-2"
      >
        <div
          className="col-12"
        >
          Question 0 - OPEN ?
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <label
            className="mt-2"
          >
            Answer
          </label>
          <div>
            <textarea
              className="form-control p-2"
              onChange={[Function]}
              rows={4}
              value="A"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
